### PR TITLE
refactor: ics29 json encoded version metadata

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Refactors the ics29 fee handshake handlers to support a JSON encoded bytestring

closes: #756 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/29-fee/keeper/keeper_test.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ func (suite *KeeperTestSuite) SetupTest() {
 	suite.chainB = suite.coordinator.GetChain(ibctesting.GetChainID(1))
 
 	path := ibctesting.NewPath(suite.chainA, suite.chainB)
-	feeTransferVersion := channeltypes.MergeChannelVersions(types.Version, transfertypes.Version)
+	feeTransferVersion := string(types.ModuleCdc.MustMarshalJSON(&types.Metadata{Version: types.Version, AppVersion: transfertypes.Version}))
```
Comment (colin-axner) [inline modules/apps/29-fee/keeper/keeper_test.go:43]: maybe we should add a `MustMarshalJSON` function to Metadata? or a function to return the marshaled metadata as a string (unsure what a good name would be)

File: modules/apps/29-fee/ibc_module.go [hunk 2]
```diff
@@ -37,22 +37,24 @@ func (im IBCModule) OnChanOpenInit(
...
+	if err := types.ModuleCdc.UnmarshalJSON([]byte(version), &versionMetadata); err != nil {
+		// Since it is valid for fee version to not be specified, the above middleware version may be for a middleware
+		// lower down in the stack. Thus, if it is not a fee version we pass the entire version string onto the underlying
+		// application.
+		return im.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID,
+			chanCap, counterparty, version)
 	}
 
+	if versionMetadata.Version != types.Version {
+		return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s got %s", types.Version, versionMetadata.Version)
```
Comment (colin-axner) [inline modules/apps/29-fee/ibc_module.go:50]:
```suggestion
		return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s, got %s", types.Version, versionMetadata.Version)
```

File: modules/apps/29-fee/ibc_module.go [hunk 3]
```diff
@@ -68,31 +70,34 @@ func (im IBCModule) OnChanOpenTry(
...
-	// application.
-	// If an invalid fee version was passed, we expect the underlying application to fail on its version negotiation.
-	if cpMwVersion == types.Version {
-		im.keeper.SetFeeEnabled(ctx, portID, channelID)
-	} else {
-		// middleware versions are not the expected version for this midddleware. Pass the full version strings along,
-		// if it not valid version for any other lower middleware, an error will be returned by base application.
-		cpAppVersion = counterpartyVersion
+	if versionMetadata.Version != types.Version {
+		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s got %s", types.Version, versionMetadata.Version)
```
Comment (colin-axner) [inline modules/apps/29-fee/ibc_module.go:82]:
```suggestion
		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s, got %s", types.Version, versionMetadata.Version)
```

**Review by colin-axner — APPROVED** [review]
Aye LGTM, nice work!

File: modules/apps/29-fee/keeper/keeper_test.go (hunk 1 above)
Comment (damiannolan) [reply modules/apps/29-fee/keeper/keeper_test.go:43]: Very much agree to be honest, this was on my mind when I was writing this. I'll think on it and add it in a follow up PR. It does start to look ugly seeing all the string conversions scattered around, I would prefer to just call some method on `Metadata` to do the job.

File: proto/ibc/applications/fee/v1/metadata.proto [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+option go_package = "github.com/cosmos/ibc-go/v3/modules/apps/29-fee/types";
+
+import "gogoproto/gogo.proto";
+
+// Metadata defines the ICS29 channel specific metadata encoded into the channel version bytestring
+// See ICS004: https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#Versioning
+message Metadata {
+  // version defines the ICS29 fee version
+  string version = 1;
```
Comment (AdityaSripal) [inline proto/ibc/applications/fee/v1/metadata.proto:13]: I would prefer we make this explicit and call this `fee_version`

**Review by AdityaSripal — CHANGES_REQUESTED** [review]
Only 1 requested change. Everything else looks good!

File: proto/ibc/applications/fee/v1/metadata.proto (hunk 4 above)
Comment (damiannolan) [reply proto/ibc/applications/fee/v1/metadata.proto]: Done!

**Review by AdityaSripal — APPROVED** [review]

**Pull request merged**